src/main/java/com/es/phoneshop/model/product/dao/impl/ArrayListProductDao.java

```diff
@@ -52,26 +48,25 @@ public void save(Product product) {
         lock.writeLock().lock();
         if (product.getId() != null) {
             Optional<Product> productToAdd = products.stream().
-                    filter(product1 -> product1.getId().equals(product.getId())).
+                    filter(currProduct -> currProduct.getId().equals(product.getId())).
                     findAny();
             if (productToAdd.isEmpty()) {
-                product.setId(currentMaxId++);
-                products.add(product);
+                addNewProduct(product,products);
             } else
                 products.set(products.indexOf(productToAdd.get()), product);
         } else {
-            product.setId(currentMaxId++);
-            products.add(product);
+            addNewProduct(product,products);
         }
         lock.writeLock().unlock();
     }
 
     @Override
     public void delete(Long id) {
         if (id == null)
-            throw new ProductNotFoundException("Parameter id is null");
+            throw new IllegalArgumentException("Parameter id is null");
         lock.writeLock().lock();
-        products.removeIf(product -> id.equals(product.getId()));
+        if(!products.removeIf(product -> id.equals(product.getId())))
+            throw new ProductNotFoundException("No element was deleted");
```

Comment:
It's better to put more meaningful message like - **"Couldn't delete product with id: <parameter value>"** 
For building such messages you can use **String.format()**